zebra-state/src/service/check.rs

```diff
@@ -27,30 +39,51 @@ pub(crate) fn block_is_contextually_valid<C>(
 ) -> Result<(), ValidateContextError>
 where
     C: IntoIterator,
-    C::Item: AsRef<Block>,
+    C::Item: Borrow<Block>,
+    C::IntoIter: ExactSizeIterator,
 {
     let finalized_tip_height = finalized_tip_height
-        .expect("finalized state must contain at least one block to use the non-finalized state");
+        .expect("finalized state must contain at least one block to do contextual validation");
     check::block_is_not_orphaned(finalized_tip_height, prepared.height)?;
 
-    let mut relevant_chain = relevant_chain.into_iter();
+    // Peek at the first block
+    let mut relevant_chain = relevant_chain.into_iter().peekable();
     let parent_block = relevant_chain
-        .next()
+        .peek()
         .expect("state must contain parent block to do contextual validation");
-    let parent_block = parent_block.as_ref();
+    let parent_block = parent_block.borrow();
     let parent_height = parent_block
         .coinbase_height()
         .expect("valid blocks have a coinbase height");
     check::height_one_more_than_parent_height(parent_height, prepared.height)?;
 
-    // TODO: validate difficulty adjustment
-    // TODO: other contextual validation design and implelentation
+    // Note: the difficulty check reads the first 28 blocks from the relevant
+    // chain iterator. If you want to use those blocks in other checks, you'll
+    // need to clone them here.
+
+    if relevant_chain.len() >= POW_AVERAGING_WINDOW + POW_MEDIAN_BLOCK_SPAN {
```

Comment:
In #1354 we want to skip this check if the chain is too short. So the calling function in this PR asserts the length of the chain, but another calling function added in #1354 does not.

Reply:
I worry that this could be a source of bugs. You placed the comment earlier noting that the iterator is modified by usage, and if other checks need access to the relevant chain they should copy the chain rather than iterating over it. If we ever accidentally violated this and popped some blocks from the relevant chain we could end up just skipping this check because the length is to short, even when it should have run.

Reply:
If we accidentally pop some blocks, the contextual checks after the pop will all fail.

With the default config, contextual difficulty checks are only run after Sapling activation, when the chain has hundreds of thousands of blocks. If we merge #1354, we'll also run the checks on the early chain, but only when `debug_contextual_verify` is set.

But I think you're right, we might need a better design here than `peek` and "please don't mangle the iterator" in a comment.

Reply:
I made a copy of the block borrows in a Vec before doing any of the contextual checks. We have to know the maximum number of blocks we might used in advance, but that seems to work for now.

See 99633017b7836855e4415c798e1f6380d05fd035